prover/crates/bin/witness_generator/src/basic_circuits.rs

```diff
@@ -432,40 +438,83 @@ async fn generate_witness(
             path,
             input.eip_4844_blobs.blobs(),
             |circuit| {
-                circuit_sender.blocking_send(circuit).unwrap();
+                circuit_sender
+                    .blocking_send(circuit)
+                    .expect("failed to send circuit from harness");
+            },
+            |a, b, c| {
+                queue_sender
+                    .blocking_send((a as u8, b, c))
+                    .expect("failed to send recursion queue from harness")
             },
-            |a, b, c| queue_sender.blocking_send((a as u8, b, c)).unwrap(),
         );
         (scheduler_witness, block_witness)
     });
 
-    let mut circuit_urls = vec![];
-    let mut recursion_urls = vec![];
-
-    let mut circuits_present = HashSet::<u8>::new();
+    let mut save_circuit_handles = vec![];
+
+    // Ordering determines how we compose the circuit proofs in Leaf Aggregation Round.
+    // Sequence is used to determine circuit ordering (the sequencing of instructions) .
+    // If the order is tampered with, proving will fail (as the proof would be computed for a different sequence of instruction).
+    let circuit_sequence = AtomicUsize::new(0);
+
+    // Future which receives circuits and saves them async.
+    let circuit_receiver_handle = async {
+        while let Some(circuit) = circuit_receiver.recv().await {
+            let sequence = circuit_sequence.fetch_add(1, Ordering::SeqCst);
+            let object_store = object_store.clone();
+            save_circuit_handles.push(tokio::task::spawn(save_circuit(
```

Comment:
Please note that `tokio::task::spawn` can be considered mostly a no-op. So this implementation _in extremum_ has even more memory consumption than two unbounded channels:

Unbound channels: any generated circuit/queue is added to the channel, but we send them to GCS _fully_ sequentially (e.g. we can't even send a circuit and a queue in parallel). So, we may have, say, 10 items in the channel, but only 1 will be in the process of sending.
Spawned tasks: any generated circuit/queue is added to the channel and immediately consumed. A new task is spawned for it. Tasks includes additional steps like serialization, which also consume memory. So, our channel will generally be empty, but we may have 10 spawned jobs at different stages (serialization/sending).

Moreover, since we will be sending data in parallel, it _may_ result in lower bandwidth for each particular task (e.g. 1gbps for 10 tasks => ~100mbps/task), so that each job is processed for longer, and the amount of jobs we have to keep in memory also grows.

It's just an observation, since the old apporach was focused on saving us from OOM, and PR description also touches that -- I think that the new approach _may_ be significantly more prone to OOM in real environments.